File: AoC2023/Day 2/day2.py
from collections import defaultdict

def readFile(path):
    games = []
    
    with open(path, 'r') as file:
        for line_num, line in enumerate(file, start=1):
            if line.startswith("Game"):
                data = line.split(":")[1].strip().split(';')
                game_subsets = []
                for subset in data:
                    cube_list = subset.split(',')
                    subset_data = [cube.strip() for cube in cube_list]
                    game_subsets.append(subset_data)
                games.append({"subsets": game_subsets, "line_num": line_num})
    return games

def possible(games, target):
    total_sum = 0
    
    for game in games:
        subsets = game["subsets"]
        total = defaultdict(int)
        
        for subset in subsets:
            for cube in subset:
                count, color = cube.split()
                total[color] = max(total[color], int(count))
                
                product = 1
                for color in total:
                    product *= total[color]
                
        print(f"Game {game['line_num']} - Total: {total} - ", product)        
        
   
        total_sum += product
        print(f"Running Total: {total_sum} (Game {game['line_num']} added)")
                
    return total_sum

path = "Advent of Code/Day 2/input.txt"

games = readFile(path)

target = {"red": 12, "green": 13, "blue": 14}

total_sum = possible(games, target)

print(total_sum)

File: AoC2015/Day 2/day2.rb
File.open("Day 2/input.txt") do |input|
    total = 0
    ribbon = 0
    input.each_line do |line|
        
        values = line.split('x')
        l = Integer(values[0])
        w = Integer(values[1])
        h = Integer(values[2])

        x = l*w
        y = w*h
        z = h*l

        s = ((x < y)?((x < z)?x : z) : ((y < z)?y : z))

        total = total + (2*x + 2*y + 2*z + s)

        #Part 2
        cf = l*w*h

        max = ((l > w)?((l > h)?l : h) : ((w > h)?w : h))

        if max == l
            ribbon += w*2 + h*2 + cf
        elsif max == w
            ribbon += l*2 + h*2 + cf
        else
            ribbon += l*2 + w*2 + cf
        end
        
    end
    puts total
    puts ribbon
end